### PR TITLE
Add Event Sourceing example for redis

### PR DESCRIPTION
**Pull request by pdambrauskas**

### CODE REVIEW

File: event-sourcing-redis/Readme.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
+# Event Sourcing with Redis
+It is a common practice to use asynchronous messaging technologies to implement communication between multiple microservices.
+It helps to have scalable, fault tolerant and decoupled architecture. In my personal project and also at work I had some experience
```
Comment (Minutis) [inline event-sourcing-redis/Readme.md]: You could elaborate a little bit more (a sentence or so) on how "It helps".

File: event-sourcing-redis/Readme.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+in implementing asynchronous communication by using Apache Kafka. Kafka is go-to solution when it comes to streaming pipelines and
```
Comment (Minutis) [inline event-sourcing-redis/Readme.md]: I would switch these two sentences places and few minor changes to the first one:

> In my personal project and also at work I had some experience in implementing asynchronous communication by using Apache Kafka. Kafka is go-to solution when it comes to streaming pipelines and publish/subscribe systems.

with

> Kafka is go-to solution when it comes to streaming pipelines and publish/subscribe systems. I have implemented asynchronous communication using Kafka several times both at work and working on personal projects.

File: event-sourcing-redis/Readme.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
+# Event Sourcing with Redis
+It is a common practice to use asynchronous messaging technologies to implement communication between multiple microservices.
+It helps to have scalable, fault tolerant and decoupled architecture. In my personal project and also at work I had some experience
+in implementing asynchronous communication by using Apache Kafka. Kafka is go-to solution when it comes to streaming pipelines and 
+publish/subscribe systems. But there are many alternatives. One of them is newest Redis datatype - Redis Streams, which came in Redis version 5.0.
```
Comment (Minutis) [inline event-sourcing-redis/Readme.md]:
```suggestion
publish/subscribe systems. But there are many alternatives. One of the latest implementations is Redis datatype - Redis Streams, which came in [version 5.0](https://redislabs.com/blog/redis-5-0-is-here/).
```

File: event-sourcing-redis/Readme.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+[Redis Streams](https://redis.io/topics/streams-intro) may look very similar to traditional Redis Pub/Sub concept, however it is quite different:
+- stream can have multiple consumers, every new entry on a stream will be delivered to every consumer (unless consumers belong to same consumer group);
+- consumed messages do not disappear, Redis stores streamed data and last consumed *id* for each consumer group,
+ so new consumer groups can consume group from the beginning.
+ 
+ In theory Redis Streams work very similar to Kafka topics, and can be used for same use cases.
+ You can find many resources on how to use Kafka for Event Sourcing ([example](https://www.confluent.io/blog/event-sourcing-cqrs-stream-processing-apache-kafka-whats-connection/)),
+ but when you look at the alternatives, choice is limited. So I decided to contribute to filling in this resource gap :).
```
Comment (Minutis) [inline event-sourcing-redis/Readme.md]: What do you mean by "choice is limited"? Are you referring that there are very few options on how to implement Event Sourcing and CQRS or that there are only few examples on how to do such a thing using other technologies than Kafka? You should be more specific.

File: event-sourcing-redis/Readme.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+ On this post I will try to describe how to use Redis Streams for [Event sourcing](https://martinfowler.com/eaaDev/EventSourcing.html)
```
Comment (Minutis) [inline event-sourcing-redis/Readme.md]:
```suggestion
 In this post, I will try to describe how to use Redis Streams for [Event sourcing](https://martinfowler.com/eaaDev/EventSourcing.html)
```

File: event-sourcing-redis/Readme.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+ & [CQRS](https://www.martinfowler.com/bliki/CQRS.html).
+ 
+ ## Redis Streams API theoretical example
+ Lets say we have a REST API endpoint for GET'ing user social media profile, with personal information,  liked pages, list of users' friends, etc.
+ For that we'll have three different Redis Streams:
+ - `user_updates` stream consists of user profile updates, every time user changes his birth date, name or other personal information, entry is published to this stream;
```
Comment (Minutis) [inline event-sourcing-redis/Readme.md]:
```suggestion
 - `user_updates` stream consists of user profile updates. Every time the user changes his birth date, name or other personal information, entry is published to this stream;
```

File: event-sourcing-redis/Readme.md [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+ - `user_activity` every time user writes a comment, reacts to some post, or does some other thing, what people do on facebook, entry is published to this stream;
```
Comment (Minutis) [inline event-sourcing-redis/Readme.md]:
```suggestion
 - `user_activity` every time user writes a comment, reacts to some post, or performs any other action (that is available in Facebook), entry is published to this stream;
```

File: event-sourcing-redis/Readme.md [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+ so new consumer groups can consume group from the beginning.
+ 
+ In theory Redis Streams work very similar to Kafka topics, and can be used for same use cases.
+ You can find many resources on how to use Kafka for Event Sourcing ([example](https://www.confluent.io/blog/event-sourcing-cqrs-stream-processing-apache-kafka-whats-connection/)),
+ but when you look at the alternatives, choice is limited. So I decided to contribute to filling in this resource gap :).
+ On this post I will try to describe how to use Redis Streams for [Event sourcing](https://martinfowler.com/eaaDev/EventSourcing.html)
+ & [CQRS](https://www.martinfowler.com/bliki/CQRS.html).
+ 
+ ## Redis Streams API theoretical example
+ Lets say we have a REST API endpoint for GET'ing user social media profile, with personal information,  liked pages, list of users' friends, etc.
```
Comment (Minutis) [inline event-sourcing-redis/Readme.md]:
```suggestion
 Lets say we have a REST API endpoint for GET'ing user social media profile, with personal information, liked pages and list of users' friends.
```

File: event-sourcing-redis/Readme.md [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+ For that we'll have three different Redis Streams:
+ - `user_updates` stream consists of user profile updates, every time user changes his birth date, name or other personal information, entry is published to this stream;
+ - `user_activity` every time user writes a comment, reacts to some post, or does some other thing, what people do on facebook, entry is published to this stream;
+ - `user_friends` every time user gets or loses a friend, entry is published to this stream.
```
Comment (Minutis) [inline event-sourcing-redis/Readme.md:27]:
```suggestion
 - `user_friends` every time user gets or loses a friend, entry is published to this stream;
```

File: event-sourcing-redis/Readme.md [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+XADD user_friends * user_id 1 command add friend_id 2
+```
+
+### Reading messages from event streams
+Now we can read all streams and build up user snapshot, using `XREAD` command:
+```shell script
+XREAD STREAMS user_updates user_activity user_friends 0 0 0
+```
+
+And you'll get output:
```
Comment (Minutis) [inline event-sourcing-redis/Readme.md]:
```suggestion
Your output should look like this:
```

File: event-sourcing-redis/Readme.md [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+            2) "1"
+            3) "action"
+            4) "add"
+            5) "friend_id"
+            6) "2"
+```
+
+### Building snapshots
+When building profile representation, you can transform and represent it in any structure you need.
+Notice, the three zeroes at the end of the command. Those are entry ids, which can be seen as consumer offsets, when you are using `XREADGROUP` command.
```
Comment (Minutis) [inline event-sourcing-redis/Readme.md]:
```suggestion
Notice, the three zeroes at the end of the command. Those are entry IDs, that are used as consumer offsets when you are using `XREADGROUP` command.
```

File: event-sourcing-redis/Readme.md [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+Entry IDs by default are represented as timestamps, with nanoseconds part after dot, so it is easy to read stream from any point in time you need.
```
Comment (Minutis) [inline event-sourcing-redis/Readme.md]:
```suggestion
Entry IDs by default are actually timestamps, with nanoseconds part after the dot. That means it is easy to read the stream from any point in time. 
```

File: event-sourcing-redis/Readme.md [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+            5) "friend_id"
+            6) "2"
+```
+
+### Building snapshots
+When building profile representation, you can transform and represent it in any structure you need.
+Notice, the three zeroes at the end of the command. Those are entry ids, which can be seen as consumer offsets, when you are using `XREADGROUP` command.
+Entry IDs by default are represented as timestamps, with nanoseconds part after dot, so it is easy to read stream from any point in time you need. 
+
+In most cases you might want to save user profile snapshots, it is not efficient to consume whole stream each time you query.
```
Comment (Minutis) [inline event-sourcing-redis/Readme.md]:
```suggestion
In most cases, it is inefficient to consume the whole stream every time you need data. It is a common practice to save your user profile snapshots in one way or another.
```

File: event-sourcing-redis/Readme.md [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+For snapshoting you can use Redis Hash data structure (or one of other Redis structures, if that makes sense for you).
+Hash structure can be created by using `HMSET` command
+(you can set multiple hash fields for aggregated data, it is skipped on this example), the suffix of Hash key is user id:
```
Comment (Minutis) [inline event-sourcing-redis/Readme.md]:
```suggestion
(you can set multiple hash fields for aggregated data, it is skipped in this example), the suffix of Hash key is user id:
```

File: event-sourcing-redis/Readme.md [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+```shell script
+HMSET user_snapshot_1 name Donal ... [field value]
```
Comment (Minutis) [inline event-sourcing-redis/Readme.md]:
```suggestion
HMSET user_snapshot_1 name Donald ... [field value]
```

File: event-sourcing-redis/Readme.md [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+```
+
+### Continuous consumption
+But you don't want to keep track of offsets, you have already consumed, right? Redis has a solution for this problem too.
```
Comment (Minutis) [inline event-sourcing-redis/Readme.md]:
```suggestion
But you don't want to keep track of offsets, you have already consumed, right? Redis has a solution to this problem too.
```

File: event-sourcing-redis/Readme.md [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+
+## Java example
+
+I've also implemented Java App to illustrate how to use Redis Streams for Event Sourcing. I chose to use lettuce library for communication with Redis server.
+You can find Source code on [Github](https://github.com/pdambrauskas/event-sourcing/event-sourcing-redis)
+
+There are three main classes:
+- EventStore - this class can be used for event publishing, and subscribing to Redis Streams;
+- SnapshotStore - this class can be used to store and retrieve snapshot objects from Redis;
+- StreamProcessor - combines EventStore and Snapshot store. You can supply multiple event handlers, which are used while building snapshots.
```
Comment (Minutis) [inline event-sourcing-redis/Readme.md]:
```suggestion
- StreamProcessor - combines EventStore and Snapshot store. You can supply multiple event handlers, which are used for building snapshots;
```

File: event-sourcing-redis/Readme.md [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+
+Whole combination of these classes can be found on RedisEventSourcingTest Unit test class.
```
Comment (Minutis) [inline event-sourcing-redis/Readme.md]:
```suggestion
The whole combination of these classes can be found in RedisEventSourcingTest unit test class.
```

**Review by Minutis — COMMENTED** [review]
Also, it would be nice to see some summary paragraph where you could share your own opinion regarding this implementation,

**Comment by pdambrauskas** [comment]
@Minutis, updated.
- Fixed some spelling & grammar mistakes you mentioned
- Elaborated more on streaming architecture advanteges
- Added some kind of conclusion :)

Can you pelaese re-review one more time ;)?

File: event-sourcing-redis/Readme.md [hunk 19]
```diff
@@ -0,0 +1,141 @@
+# Event Sourcing with Redis
+It is a common practice to use asynchronous messaging technologies to implement communication between multiple microservices.
+It helps to have scalable, fault tolerant and decoupled architecture. Kafka is go-to solution when it comes to streaming pipelines and publish/subscribe systems.
```
Comment (Minutis) [inline event-sourcing-redis/Readme.md]:
```suggestion
It provides scalable, fault tolerant and decoupled architecture. Kafka is go-to solution when it comes to streaming pipelines and publish/subscribe systems.
```

File: event-sourcing-redis/Readme.md [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+You can easily scale microservice architecture by running multiple instances of your microservice in same consumer group, easily add new consumers without modifying existing communication-related code,
+take advantage of persistence of Kafka topics by re-consuming them any time you need and have many more advantages over traditional http communication protocol.
```
Comment (Minutis) [inline event-sourcing-redis/Readme.md]: Change this sentence to bullet points. I.E. 
Main advantages over traditional HTTP communication protocol:
- You can easily scale microservice architecture by running multiple instances of your microservice in same consumer group.
- Easily add new consumers without modifying existing communication-related code.
- Take advantage of persistence of Kafka topics by re-consuming them any time you need

File: event-sourcing-redis/Readme.md [hunk 21]
```diff
@@ -0,0 +1,141 @@
+# Event Sourcing with Redis
+It is a common practice to use asynchronous messaging technologies to implement communication between multiple microservices.
+It helps to have scalable, fault tolerant and decoupled architecture. Kafka is go-to solution when it comes to streaming pipelines and publish/subscribe systems.
+You can easily scale microservice architecture by running multiple instances of your microservice in same consumer group, easily add new consumers without modifying existing communication-related code,
+take advantage of persistence of Kafka topics by re-consuming them any time you need and have many more advantages over traditional http communication protocol. 
+I have implemented asynchronous communication using Kafka several times both at work and working on personal projects
+des. But there are many alternatives. One of the latest implementations is Redis datatype - Redis Streams, which came in Redis [version 5.0](https://redislabs.com/blog/redis-5-0-is-here/).
```
Comment (Minutis) [inline event-sourcing-redis/Readme.md]: What is `des`?

File: event-sourcing-redis/Readme.md [hunk 22]
```diff
@@ -0,0 +1,141 @@
...
+         2) 1) "user_id"
+            2) "1"
+            3) "action"
+            4) "add"
+            5) "friend_id"
+            6) "2"
+```
+
+### Building snapshots
+When building profile representation, you can transform and represent it in any structure you need.
```
Comment (Minutis) [inline event-sourcing-redis/Readme.md]:
```suggestion
When building profile representation, you can transform and reflect it in any structure you need.
```

File: event-sourcing-redis/Readme.md [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+            process_message(message)
+            XACK message.stream snapshotter message.id
+        END
+    END
+END
+```
+
+## Java example
+
+I've also implemented Java App to illustrate how to use Redis Streams for Event Sourcing. I chose to use lettuce library for communication with Redis server.
```
Comment (Minutis) [inline event-sourcing-redis/Readme.md]:
```suggestion
I've also implemented Java App to illustrate how to use Redis Streams for Event Sourcing. I chose to use [lettuce](https://github.com/lettuce-io/lettuce-core) library for communication with Redis server.
```

File: event-sourcing-redis/Readme.md [hunk 24]
```diff
@@ -0,0 +1,141 @@
...
+You can find Source code on [Github](https://github.com/pdambrauskas/event-sourcing/event-sourcing-redis)
```
Comment (Minutis) [inline event-sourcing-redis/Readme.md]:
```suggestion
Source code of my implementation can be found on [Github](https://github.com/pdambrauskas/event-sourcing/event-sourcing-redis)
```

File: event-sourcing-redis/Readme.md [hunk 25]
```diff
@@ -0,0 +1,141 @@
...
+There are three main classes:
+- EventStore - this class can be used for event publishing, and subscribing to Redis Streams.
+- SnapshotStore - this class can be used to store and retrieve snapshot objects from Redis.
+- StreamProcessor - combines EventStore and Snapshot store. You can supply multiple event handlers, which are used for building snapshots.
+
+The whole combination of these classes can be found on RedisEventSourcingTest Unit test class.
+
+## Concludion
+While Redis Streams is relatively new concept, it is heavily inspired by Apache Kafka, and has many overlapping features. Also we mustn't forget that
+Redis has many more data structures and features, which can be used alongside Redis streams (one of which we so on snapshoting example in this post).
```
Comment (Minutis) [inline event-sourcing-redis/Readme.md]:
```suggestion
Redis has many more data structures and features, which can be used alongside Redis streams (one of which we used for the snapshotting example in this post).
```

File: event-sourcing-redis/Readme.md (hunk 19 above)
Comment (Minutis) [reply event-sourcing-redis/Readme.md]: If I understood the idea, this is more appropriate and leaves no place for misinterpretation.

**Pull request merged**